egui_demo_lib/src/apps/fractal_clock.rs

```diff
@@ -200,5 +206,33 @@ impl FractalClock {
 
             std::mem::swap(&mut nodes, &mut new_nodes);
         }
+        self.line_count = shapes.len();
+        painter.extend(shapes);
+    }
+
+    fn clip_line(&self, input: [Pos2; 2], rect: Rect) -> bool {
+        let mask_a: u8 = self.clip_line_mask(input[0], rect);
+        let mask_b: u8 = self.clip_line_mask(input[1], rect);
+        (mask_a & mask_b) == 0
+    }
+    fn clip_line_mask(&self, p: Pos2, rect: Rect) -> u8 {
+        let mut mask: u8 = 0;
+        let top = rect.top();
+        let btm = rect.bottom();
+        let lft = rect.left();
+        let rgt = rect.right();
+        if p.y < top {
+            mask |= 0b1000;
+        }
+        if p.y > btm {
+            mask |= 0b0100;
+        }
+        if p.x < lft {
+            mask |= 0b0010;
+        }
+        if p.x > rgt {
+            mask |= 0b0001;
+        }
+        mask
```

Comment:
```suggestion
```

I rather not make this longer or complex. This is a demo of painting with egui, not a demo of performance.

Also `clip_line` is misleading: this is not doing [line clipping](https://en.wikipedia.org/wiki/Line_clipping) but [view culling](https://en.wikipedia.org/wiki/Hidden-surface_determination#Viewing-frustum_culling).

 If you really want this (and it really improves perf a lot) then please make it dead simple with `rect.instersects(Rect::from_points(input))`

Reply:
Yes it is more brief.